ethexe/tx-pool/src/validation.rs

```diff
@@ -0,0 +1,373 @@
+        self
+    }
+
+    pub(crate) fn with_mortality_check(mut self) -> Self {
+        self.mortality_check = true;
+        self
+    }
+
+    pub(crate) fn with_uniqueness_check(mut self) -> Self {
+        self.uniqueness_check = true;
+        self
+    }
+
+    pub(crate) fn with_executable_tx_check(mut self, sender: TxPoolOutputTaskSender<Tx>) -> Self {
+        self.executable_tx_check = Some(sender);
+        self
+    }
+}
+
+impl<Tx> TxValidator<Tx>
+where
+    Tx: TransactionTrait + Send + Sync + 'static,
+{
+    /// Runs all sync and async validators for the transaction.
+    pub(crate) async fn full_validate(self) -> Result<Tx> {
+        self.validate_inner()?;
+        self.async_validate().await
+    }
+
+    /// Runs all stateful and stateless sync validators for the transaction.
+    pub(crate) fn validate(self) -> Result<Tx> {
+        self.validate_inner()?;
+
+        Ok(self.transaction)
+    }
+
+    fn validate_inner(&self) -> Result<()> {
+        if self.signature_check {
+            self.check_signature()?;
+        }
+
+        if self.mortality_check {
+            self.check_mortality()?;
+        }
+
+        if self.uniqueness_check {
+            self.check_uniqueness()?;
+        }
+
+        Ok(())
+    }
+
+    /// Runs all async validators for the transaction.
+    pub(crate) async fn async_validate(self) -> Result<Tx> {
+        if self.executable_tx_check.is_some() {
+            self.check_is_executable_tx().await?;
+        }
+
+        Ok(self.transaction)
+    }
+
+    /// Validates transaction signature.
+    fn check_signature(&self) -> Result<()> {
+        let tx_digest = self.transaction.encode().to_digest();
+        let signature = self.transaction.signature()?;
+
+        signature.verify_with_public_key_recover(tx_digest)
+    }
+
+    /// Validates transaction mortality.
+    ///
+    /// Basically checks that transaction reference block hash is within the recent blocks window.
+    fn check_mortality(&self) -> Result<()> {
+        let block_hash = self.transaction.reference_block_hash();
+
+        if self.db.check_within_recent_blocks(block_hash) {
+            Ok(())
+        } else {
+            Err(anyhow!("Transaction out of recent blocks window"))
+        }
+    }
+
+    /// Validates transaction uniqueness.
+    ///
+    /// Basically checks that transaction is not already in the database.
+    fn check_uniqueness(&self) -> Result<()> {
+        let tx_hash = self.transaction.tx_hash();
+
+        if self.db.validated_transaction(tx_hash).is_none() {
+            Ok(())
+        } else {
+            Err(anyhow!("Transaction already exists"))
+        }
+    }
+
+    /// Validates if transaction is executable.
+    ///
+    /// Basically sends the transaction to the external transaction pool service
+    /// to check if it is executable.
+    async fn check_is_executable_tx(&self) -> Result<()> {
+        let Some(task_sender) = self.executable_tx_check.as_ref() else {
```

Comment:
review: instead of sending task to the main service, instantiate `Processor` and do the job here

Reply:
done